File: lab11/lab11-1.py
class Solution:
    def islandPerimeter(self, grid):
        rows, cols = len(grid), len(grid[0])
        visited = set()

        def dfs(r, c):
            
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == 0:
                return 1
            if (r, c) in visited:
                return 0
            
            visited.add((r, c))
            perimeter = 0
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                perimeter += dfs(r + dr, c + dc)
            return perimeter

        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    return dfs(r, c)

File: lab11/lab11-1.c
#include <stdio.h>

#define MAX_ROWS 100
#define MAX_COLS 100

int rows, cols;
int grid[MAX_ROWS][MAX_COLS];
int visited[MAX_ROWS][MAX_COLS];

int dfs(int r, int c) {
    if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == 0) {
        return 1;
    }
    if (visited[r][c]) {
        return 0;
    }

    visited[r][c] = 1;
    int perimeter = 0;
    int directions[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    for (int i = 0; i < 4; i++) {
        int dr = directions[i][0];
        int dc = directions[i][1];
        perimeter += dfs(r + dr, c + dc);
    }
    return perimeter;
}

int islandPerimeter(int grid[][MAX_COLS], int r, int c) {
    rows = r;
    cols = c;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            visited[i][j] = 0;
        }
    }
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            if (grid[i][j] == 1) {
                return dfs(i, j);
            }
        }
    }

    return 0;  
}

int main() {
    int grid[4][4] = {
        {0, 1, 0, 0},
        {1, 1, 0, 0},
        {0, 1, 1, 0},
        {0, 0, 0, 1}
    };
    int rows = 4, cols = 4;
    printf("Island perimeter: %d\n", islandPerimeter(grid, rows, cols));
    return 0;
}